### ISSUE TITLE
Pre-PR Feedback: Custom logits processing / token filtering

### ISSUE THREAD

**Issue by noamgat** [huggingface/text-generation-inference#1269]
### Feature request

Text-generation-inference is a leading inference server for high performance model serving.
However, some use cases require a custom logits warper in order to correctly function. For this reason, huggingface API added things like `prefix_allowed_tokens_fn` parameter to `model.generate()`.

How can we add this kind of functionality to text-generation-inference?
There are two possible approaches:
1) Doing this at the server level - when the server is set up. This means that all of the requests will use the custom logits processor.
2) Doing this at the request level - adding a parameter (or several) to [NextTokenChooserParameters](https://github.com/huggingface/text-generation-inference/blob/8acdc1fae79053ae08f7cf809e1d6331f3a6a8c8/proto/generate.proto#L53C23-L53C23) and running a custom logics processor on a per request basis.

Which approach is more inline with the philosophy of the engine?
Would it be better to have a generic extendible interface, or maybe hard-code specific logits processors and interface directly with them in the parameters?

### Motivation

I am the author of [lm-format-enforcer](https://github.com/noamgat/lm-format-enforcer), a library that enforces JSON Schema / Regex format of LLM output.

It is similar to guidance / outlines / jsonformer in some ways, but is unique in that it is much easier to integrate into other pipelines (already integrated in transformers, llama-cpp-python, vllm, haystack, llama-index, langchain) and is the only one that supports batched generation, making it ideal for high performance scenarios like this server is intended to host.

The library's users have asked about integration with this inference engine, and I think there is an audience for these features, especially now that OpenAI released JSON mode, which this plugin also enables.

### Your contribution

I am willing to write the PR, if we agree on the design principles that will make it be accepted.

**Comment by mjsteele12** [comment]
I am just a user, but I would vastly prefer option #2. I use TGI heavily for data processing, and being able to support multiple pipelines without needing multiple servers/restarting the server in order to change formats would be _immensely_ helpful. I had not come across your library before, it looks awesome btw, thank you for the hard work!!
